Refactoring before implementing on-comments PR

chmouel opened the pull request:
refactor the code to use the TriggerType type instead of options which
was all over the place

Move all comments logic to a separate package
    
There is still plenty to be done but trying to do it step by step as this code is quite complex and need to be tested properly.
    

Signed-off-by: Chmouel Boudjnah <chmouel@redhat.com># Changes <!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

<!-- Describe your changes here- ideally you can get that description straight from
your descriptive commit message(s)! -->

# Submitter Checklist

- [ ] 📝 Please ensure your commit message is clear and informative. For guidance on crafting effective commit messages, refer to the How to write a git commit message guide. We prefer the commit message to be included in the PR body itself rather than a link to an external website (ie: Jira ticket).

- [ ] ♽ Before submitting a PR, run make test lint to avoid unnecessary CI processing. For an even more efficient workflow, consider installing [pre-commit](https://pre-commit.com/) and running pre-commit install in the root of this repository.

- [ ] ✨ We use linters to maintain clean and consistent code. Please ensure you've run make lint before submitting a PR. Some linters offer a --fix mode, which can be executed with the command make fix-linters (ensure [markdownlint](https://github.com/DavidAnson/markdownlint) and [golangci-lint](https://github.com/golangci/golangci-lint) tools are installed first).

- [ ] 📖 If you're introducing a user-facing feature or changing existing behavior, please ensure it's properly documented.

- [ ] 🧪 While 100% coverage isn't a requirement, we encourage unit tests for any code changes where possible.

- [ ] 🎁 If feasible, please check if an end-to-end test can be added. See [README](https://github.com/openshift-pipelines/pipelines-as-code/blob/main/test/README.md) for more details.

- [ ] 🔎 If there's any flakiness in the CI tests, don't necessarily ignore it. It's better to address the issue before merging, or provide a valid reason to bypass it if fixing isn't possible (e.g., token rate limitations).

chmouel committed:
Move everything to use the TriggerType type

refactor the code to use the TriggerType type instead of options which
was all over the place

Signed-off-by: Chmouel Boudjnah <chmouel@redhat.com>

chmouel committed:
Move all comments logic to a separate package

There is still plenty to be done but trying to do it step by step as
this code is quite complex and need to be tested properly.

Signed-off-by: Chmouel Boudjnah <chmouel@redhat.com>

piyush-garg commented on pkg/opscomments/comments.go:
```diff
@@ -0,0 +1,171 @@
+		eventType == OkToTestCommentEventType.String()
+}
+
+func IsAnyOpsEventType(eventType string) bool {
```
this and above function are duplicate

chmouel committed:
Reorganize the GitHub test

we now use a struct and function attached to it instead of passing
everything around all the times

this will be more readable and easier to maintain

Signed-off-by: Chmouel Boudjnah <chmouel@redhat.com>

savitaashture commented on pkg/opscomments/comments.go:
```diff
@@ -0,0 +1,161 @@
+	NoOpsCommentEventType        = EventType("no-ops-comment")
+	TestCommentEventType         = EventType("test-comment")
+	RetestSingleCommentEventType = EventType("retest-comment")
+	RetestAllCommentEventType    = EventType("retest-all-comment")
```
I see SingleCommentEventType and  AllCommentEventType for `retest` & `cancel` but not for `test`
any reason ?
asking because `test` also works similar to `retest`

chmouel replied:
issuing a single /test doesn't issue a /retest of all pipelines, only /retest does afaik,

chmouel replied:
ah yeah it does, i think the initial design we said that only retest retest all (like prow) and the previous code says it as much by not defining a TestAll Regexp, i guess it's a side effect we have to live forever....

savitaashture replied:
btw what is the difference between `retest` and `test` ? because both comment will trigger the CI 
So i am not able to see difference between those 2  :thinking:

chmouel replied:
it's a categorization to know where things come from if the pr got started with /retest it would be set in the eventtype or /test

I was thinking to remove it and  remove the retest type all together but i think it's fine to keep it since that makes things less surprising

savitaashture commented on pkg/opscomments/comments_test.go:
```diff
@@ -0,0 +1,627 @@
+			wantTestPr: "prname",
+		},
+		{
+			name:       "no-ops-commenttest event type",
```
```suggestion
			name:       "test single event type",
```

savitaashture commented on pkg/policy/policy.go:
```diff
@@ -39,12 +40,12 @@ func (p *Policy) checkAllowed(ctx context.Context, tType info.TriggerType) (Resu
-	// NOTE: not supported yet, will imp if it gets requested and reasonable to implement
-	case info.TriggerTypePush, info.TriggerTypeCancel, info.TriggerTypeCheckSuiteRerequested, info.TriggerTypeCheckRunRerequested:
+		// NOTE: not supported yet, will imp if it gets requested and reasonable to implement
+	case triggertype.Push, triggertype.Cancel, triggertype.CheckSuiteRerequested, triggertype.CheckRunRerequested, triggertype.Incoming:
```
ah incoming was not there earlier 
:+1:

savitaashture commented on pkg/provider/github/github.go:
```diff
@@ -432,7 +433,9 @@ func (v *Provider) getPullRequest(ctx context.Context, runevent *info.Event) (*i
 	runevent.HeadURL = pr.GetHead().GetRepo().GetHTMLURL()
 	runevent.BaseURL = pr.GetBase().GetRepo().GetHTMLURL()
-	runevent.EventType = "pull_request"
+	if runevent.EventType == "" {
+		runevent.EventType = triggertype.PullRequest.String()
+	}
```
for getPullRequest I believe  evnteType is always `pull_request`

so do we need if validate `""` :thinking:

chmouel replied:
yes, because we set otherwise "on-comment" for the on-comment pr i am following next

savitaashture commented on pkg/opscomments/comments.go:
```diff
@@ -0,0 +1,161 @@
+	TestCommentEventType         = EventType("test-comment")
+	RetestSingleCommentEventType = EventType("retest-comment")
+	RetestAllCommentEventType    = EventType("retest-all-comment")
+	OnCommentEventType           = EventType("on-comment")
```
Can you add a comment for `on-comment` purpose as well as some unit test how user can pass `on-comment` ?

chmouel replied:
yeah it's following PR, it was too annoying to break down all refactoring and implementations.. but that's fine since it's coming down..

chmouel committed:
Fix all revive errors

For whatever reasons the revive linter is now working and reports
errors, so had the very boring and repetitive task of fixing all of
them.... I guess it's how i send my Friday mornings.

chmouel committed:
Use local binary for shellcheck

This was using docker previously but we can use the local binary
directly and we don't need a docker running......

Correct make lint to run on all files types no matter what i guess

Use local binary for shellcheck

This was using docker previously but we can use the local binary
directly and we don't need a docker running......

Correct make lint to run on all files types no matter what i guess

Signed-off-by: Chmouel Boudjnah <chmouel@redhat.com>

chmouel commented:
I add the testSingle and testAll type like retest in latest iterations.. and moved the on-comment stuff out of here.. 

there is a few CI fixes as well in there, seems like revive linter started to work with go1.22 or latest golangci-lint on my laptop even tho it wasn't showing the errors previously or on CI so i had to correct all of them (it's a pain), feel free to review individual commits to make it easier to understand

savitaashture commented on test/pkg/wait/wait.go:
```diff
@@ -36,14 +37,14 @@ func UntilMinPRAppeared(ctx context.Context, clients clients.Clients, opts Opts,
 	ctx, cancel := context.WithTimeout(ctx, opts.PollTimeout)
 	defer cancel()
-	return kubeinteraction.PollImmediateWithContext(ctx, opts.PollTimeout, func() (bool, error) {
+	var repo *pacv1alpha1.Repository
```
I did not see the usage of this `Repository` return :thinking: 
any reason for returning repo ?

chmouel replied:
i am using it in the follow up pr in here https://github.com/chmouel/pipelines-as-code/blob/2c09d026be3fd80ab456b18bee77b37c172c93ba/test/gitea_test.go#L797-L799

savitaashture replied:
Ah okok

piyush-garg commented:
lgtm
